### PR TITLE
Run prettier instance in `worker_threads`

### PR DESCRIPTION
**Pull request by sosukesuzuki**
This Pull Request runs instances of Prettier on `worker_threads`.

Prettier v3 depends on `import()` for loading configs and plugins. However, `import()` cannot be executed from the client of the VSCode extension, because the VSCode extension is run with `vm.Script`. For details, please see https://github.com/microsoft/vscode/issues/130367.

We initially tried to solve this problem by introducing a language server, much like vscode-eslint. The language server can execute `import()`. Therefore, by running instances of Prettier on the language server, we should be able to resolve this issue.

I created a prototype of this. However, I determined that it is difficult to introduce a language server while maintaining the current behavior of prettier-vscode as much as possible.

- https://github.com/sosukesuzuki/prettier-language-server
- https://github.com/sosukesuzuki/prettier-language-server-vscode-client

Therefore, I decided to take another approach.

According to @liuxingbaoyu's comment (https://github.com/prettier/prettier-vscode/pull/2947#issuecomment-1575066828), VSCode extensions can use `worker_threads` and a Worker can execute `import()`.

Thus, this PR makes it so that instances of Prettier are run on `Worker` of `worker_threads`.

However, the current default instance of prettier-vscode is 2.x, which does not depend on `import()` and does not need to be run on `Worker`. Therefore, only those loaded from `node_modules` will be executed on `Worker`.

- [x] Run tests
- [x] Update the [`CHANGELOG.md`][1] with a summary of your changes

[1]: https://github.com/prettier/prettier-vscode/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by sosukesuzuki** [comment]
Seems to have passed all tests except tests for plugins.

**Comment by fisker** [comment]
I've been very busy recently, and I'm not familiar with VSCode, but I'll try to find some time to take a look.

**Comment by sosukesuzuki** [comment]
I have a package running `vsce package` on my local machine (MacOS) that works fine, but downloading the package workflow artifact from GitHub Actions doesn't work...

**Comment by sosukesuzuki** [comment]
I have checked that the vsix file generated by https://github.com/prettier/prettier-vscode/actions/runs/5230728039/jobs/9444467839 works properly on my local VSCode(MacOS).

If possible, download the artifact with the following command and load it into VSCode to see if it works:

```
gh run download 5230728039
```

**Comment by liuxingbaoyu** [comment]
I tried it and it works fine. Amazing, thanks!

**Comment by sosukesuzuki** [comment]
@fisker @ntotten Hi, could you review this PR?

File: src/worker/prettier-instance-worker.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+              parentPort.postMessage({
+                type,
+                payload: { result: error, id, isError: true },
+              });
+            }
+          },
+          (reason) => {
+            parentPort.postMessage({
+              type,
+              payload: { resuslt: reason, id, isError: true },
```
Comment (fisker) [inline src/worker/prettier-instance-worker.js:84]: Typo?

File: src/worker/prettier-instance-worker.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+          result = { languages: result.languages };
+        }
+        parentPort.postMessage({
+          type,
+          payload: { result, id, isError: false },
+        });
+      } catch (error) {
+        parentPort.postMessage({
+          type,
+          payload: { result: error, id, isError: true },
```
Comment (fisker) [inline src/worker/prettier-instance-worker.js:102]: As I understant extra properties on `error` can't be cloned, for example parse error may have `codeFrame` assigned to error. But I'm not sure if we cares. 

https://github.com/prettier/prettier-synchronized/blob/97d9ba7abec8ea7f8de99ad182b6354f8739fa67/index.cjs#L44-L50

File: src/PrettierWorkerInstance.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    options?: PrettierOptions
+  ): Promise<string> {
+    const result = this.callMethod("format", [source, options]);
+    return result;
+  }
+
+  public async getSupportInfo(): Promise<{
+    languages: PrettierSupportLanguage[];
+  }> {
+    const result = await this.callMethod("getSupportInfo", []);
```
Comment (fisker) [inline src/PrettierWorkerInstance.ts:87]: Why these two method, one `await`ed, but another one not.

File: src/PrettierWorkerInstance.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  private importResolver: {
+    resolve: (version: string) => void;
+    reject: (version: string) => void;
+  } | null = null;
+
+  private callMethodResolvers: {
+    id: number;
+    resolve: (value: unknown) => void;
+    reject: (value: unknown) => void;
+  }[] = [];
```
Comment (fisker) [inline src/PrettierWorkerInstance.ts:34]: Maybe we can use `Map` or `Object` to store? so we can get it easier.

```
callMethodResolvers.findIndex
callMethodResolvers.splice
```

to

```
callMethodResolvers[id]
delete callMethodResolvers[id]
```

File: src/ModuleResolver.ts [hunk 5]
```diff
@@ -178,31 +181,23 @@ export class ModuleResolver implements ModuleResolverInterface {
...
-      const isPrettierInstance = !!moduleInstance.format;
-      const isValidVersion =
-        moduleInstance.version &&
-        !!moduleInstance.getSupportInfo &&
-        !!moduleInstance.getFileInfo &&
-        !!moduleInstance.resolveConfig &&
-        semver.gte(moduleInstance.version, minPrettierVersion);
-
-      if (!isPrettierInstance && prettierPath) {
+      const version = await moduleInstance.import();
```
Comment (fisker) [inline src/ModuleResolver.ts:184]: We can't really access all properties of the module. I'd do 

```
await moduleInstance.get('version')
```

File: src/worker/prettier-instance-worker.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    case "callMethod": {
+      const { modulePath, methodName, methodArgs, id } = payload;
+      const prettierInstance = path2ModuleCache.get(modulePath);
+      if (!prettierInstance) {
+        parentPort.postMessage({
+          type,
+          payload: {
+            result: new Error(
+              "Cannot load prettier instance from cache. Please call import() before call callMethod."
```
Comment (fisker) [inline src/worker/prettier-instance-worker.js:45]: We have `modulePath` in `PrettierWorkerInstance` we can pass it every time or set it once, and `import` it first here.

File: src/ModuleResolver.ts (hunk 5 above)
Comment (sosukesuzuki) [reply src/ModuleResolver.ts:184]: We don't really need to access any properties other than version, so I think we can leave it as is.

File: src/worker/prettier-instance-worker.js (hunk 2 above)
Comment (ntotten) [reply src/worker/prettier-instance-worker.js:97]: Would it be better to serialize the error using serialize-error? https://www.npmjs.com/package/serialize-error

Comment (sosukesuzuki) [reply src/worker/prettier-instance-worker.js:97]: serialize-error is a pure ESM package, so it is difficult to use from VSCode clients...

Comment (sosukesuzuki) [reply src/worker/prettier-instance-worker.js]: Add simple serializer [901d664](https://github.com/prettier/prettier-vscode/pull/3016/commits/901d664801d8259f230eceb5e4db3723ab88791b)

**Comment by liuxingbaoyu** [comment]
Now the error message is not as pretty as the old one, maybe we can improve it!
image
vs
image

**Comment by fisker** [comment]
This solution should work.

https://github.com/prettier/prettier-synchronized/blob/97d9ba7abec8ea7f8de99ad182b6354f8739fa67/worker.js#L11-L14
https://github.com/prettier/prettier-synchronized/blob/97d9ba7abec8ea7f8de99ad182b6354f8739fa67/index.cjs#L48-L50

**Comment by liuxingbaoyu** [comment]
To be honest I don't understand why it works. From what I understand it is using a different name.🤔
image

**Comment by sosukesuzuki** [comment]
Just doing Object.assign was enough...?

**Comment by sosukesuzuki** [comment]
No, just rejecting the error object passed from the worker produced the same output

**Comment by sosukesuzuki** [comment]
Error serialization may not be necessary here.  We can add it if we need it.

**Comment by JounQin** [comment]
I don't quite understand, `await import()` should has been supported, but as we're using TypeScript, it maybe compiled into `require` again.

But we can use the following hack:

https://github.com/just-jeb/angular-builders/blob/e8d063cd16346514d55ec6e25d41c184a999ca42/packages/custom-webpack/src/utils.ts#L68-L70

**Comment by JounQin** [comment]
Besides, the current solution does not consider yarn PnP support, [`synckit`](https://github.com/un-ts/synckit) may also help.

**Pull request merged**